src/CoreWCF.Primitives/src/CoreWCF/Dispatcher/SecurityImpersonationBehavior.cs

```diff
@@ -329,7 +329,14 @@ private ServiceSecurityContext GetAndCacheSecurityContext(MessageRpc rpc)
                 SecurityMessageProperty securityContextProperty = rpc.Request.Properties.Security;
                 if (securityContextProperty == null)
                 {
-                    securityContext = null; // SecurityContext.Anonymous
+                    if (rpc.Operation.Impersonation == ImpersonationOption.Required)
+                    {
```

Comment:
This code was correct. If you don't have a SecurityMessageProperty, then that means it wasn't populated when the client authenticated. What this code now does is assume the identity that the service was launched with and impersonate that user. This means if a method is attributed to require impersonation and the user didn't authenticate (could have multiple service endpoints with one having no authentication), they can now call the method, and worse, they will execute it with the credentials that the service is running as. If the Security property is null, it means it wasn't populated earlier by the authentication code, and that needs to be fixed. In this case, that means the ServicemodelHttpMiddleware.

Reply:
Thanks, Matt. That makes sense- I will take a look ServiceModelHttpMiddleware